### DISCUSSION TITLE
This is cool!

### DISCUSSION THREAD

**Discussion by KFeye** [iferres/long16S#11] [Q&A]
Hello!

So, I am slightly obsessed with the idea of this analysis tool, so I apologize if I ask too many questions.  I am trying to run the program (MacOS) and I have been using the barcode kit (96 expansion) and am trying very hard to get to a spot where I have taxa bar plots and diversity analysis. 

I demultiplexed and trimmed in guppy.  I have 96 folders with fastq.gz files (multiple per barcode).  How would I read that data in and do I need to pre-process it any other way (so unzip everything and put it in one fast.gz file??  Also, does it make sense to go to google cloud?  Is there a metadata file I need to create? 

This is truly great work.  Thank you!

**Comment by iferres** [comment]
Hi @KFeye, thanks for your words haha. Very funny the "obsessed" part. We are glad you find it useful.

Idk your experience with nextflow and docker/singularity, but the pipeline should work on your data. Are you having any problem with that?

May be use `--fq */*.fastq.gz` to grab files within sub-folders, that should work.

 If it's already demultiplexed, pass the `--isDemultiplexed` to avoid running Porechop. The pipeline will consider each file as a different barcode and will process them separately. 

There's also an issue with the conda version of Nanoplot, which makes the pipeline fail. Use `--noNanoplot` if that happens.

**Reply by KFeye** [reply]
I am having issues with it.  I have docker.  It has been claiming that the data does not exist? 

So for the --fq */*.fastq.qz, would I need to direct that to the directory that has the subdirectories?  I tried zipping it all and it just kind of crashes the computer.  Each directory that correlates to a barcode can have multiple .gz files.  It is demultiplexed and trimmed (barcode and adapter).  

Thank you for your help.  I've been dying to find a tool like this.  My background is in QIIME2, so I am new to this.

**Reply by iferres** [reply]
Yes. If, for instance, you have your reads at `./relative/path/to/reads` you would use `--fq ./relative/path/to/reads/*/*.fastq.gz`. 

Oh, for that data organization I'm afraid you will have to re order it. The easiest way I think it would be to pass the original fastq files, without demultiplexing them, and leave that to `porefile`. Alternatively you would have to concatenate all the files for each barcode in a single file (one per barcode), and name it in a way you could use a regex to grab those and not the rest.

If the reads are in plain text, you don't have to gzip them. Porefile will detect that. If they are, it will automatically gunzip them.

**Reply by KFeye** [reply]
Thank you!  Can we start with fast5?  

Honestly, this great!  Thank you thank you!!!

**Reply by iferres** [reply]
No, not from fast5. ONT basecallers have proprietary code, so I can't put it into a docker container and distribute it.

**Reply by KFeye** [reply]
I am re-running guppy as I got rid of my original files (multiplexed fastq).  That should be done sometime tomorrow or Thursday.  I already determined that fast5 won't work.  So, if it is okay, can I circle back when I try this again Thursday?   

Again, thank you x 1000000000000!  This is awesome!

**Reply by iferres** [reply]
No problem!

**Reply by KFeye** [reply]
Good Morning!  I now have .fastq files.  My directory is located on my desktop and I just have a few files in there to test the system out.  My command is as follows, as is my error:

(base) Drs-iMac:Desktop dr$ nextflow run microgenlab/porefile --fq '~/Desktop/blueplatefiles/*/*.fastq' --minimap2
N E X T F L O W  ~  version 20.10.0
Launching `microgenlab/porefile` [curious_poitras] - revision: 69606b1621 [master]
.____   __  ____  ____  ____  __  __    ____ 
(  _ \ /  \(  _ \(  __)(  __)(  )(  )  (  __)
 ) __/(  O ))   / ) _)  ) _)  )( / (_/\ ) _) 
(__)   \__/(__\_)(____)(__)  (__)\____/(____)
---------------------------------------------
... A full-length 16S profiling Pipeline ....
---------------------------------------------
[-        ] process > SetSilva:downloadFasta         [  0%] 0 of 1
executor >  local (1)
[-        ] process > SetSilva:downloadFasta         [  0%] 0 of 1
[-        ] process > SetSilva:downloadMeganSynMap   -
[-        ] process > Demultiplex:Concatenate        -
[-        ] process > Demultiplex:Porechop           -
[-        ] process > QFilt:NanoFilt                 -
[-        ] process > QFilt:AutoMap                  -
[-        ] process > QFilt:Yacrd                    -
[-        ] process > QCheck:NanoPlotRaw             -
[-        ] process > QCheck:NanoPlotFilt            -
[-        ] process > QCheck:SummaryTable            -
[-        ] process > Fastq2Fasta                    -
[-        ] process > Minimap2Workflow:MakeMinimapDB -
[-        ] process > Minimap2Workflow:Minimap2      -
[-        ] process > Minimap2Workflow:Sam2Rma       -
[-        ] process > Minimap2Workflow:Rma2Info      -
[-        ] process > MergeResults                   -
No files match pattern `*/*.fastq` at path: ~/Desktop/blueplatefiles/


Thank you in advance for your help and continued awesomeness!

**Reply by KFeye** [reply]
I finally got somewhere with this...but still have an error...

(base) Drs-iMac:Desktop dr$ nextflow run microgenlab/porefile --fq 'blueplatefiles/' --minimap2
N E X T F L O W  ~  version 20.10.0
Launching `microgenlab/porefile` [infallible_goldstine] - revision: 69606b1621 [master]
.____   __  ____  ____  ____  __  __    ____ 
(  _ \ /  \(  _ \(  __)(  __)(  )(  )  (  __)
 ) __/(  O ))   / ) _)  ) _)  )( / (_/\ ) _) 
(__)   \__/(__\_)(____)(__)  (__)\____/(____)
---------------------------------------------
... A full-length 16S profiling Pipeline ....
---------------------------------------------
WARN: Killing pending tasks (2)
executor >  local (3)
[-        ] process > SetSilva:downloadFasta         -
[-        ] process > SetSilva:downloadMeganSynMap   -
[ed/83f3fa] process > Demultiplex:Concatenate        [100%] 1 of 1, failed: 1 ✘
[-        ] process > Demultiplex:Porechop           -
[-        ] process > QFilt:NanoFilt                 -
[-        ] process > QFilt:AutoMap                  -
[-        ] process > QFilt:Yacrd                    -
[-        ] process > QCheck:NanoPlotRaw             -
[-        ] process > QCheck:NanoPlotFilt            -
[-        ] process > QCheck:SummaryTable            -
[-        ] process > Fastq2Fasta                    -
[-        ] process > Minimap2Workflow:MakeMinimapDB -
[-        ] process > Minimap2Workflow:Minimap2      -
[-        ] process > Minimap2Workflow:Sam2Rma       -
[-        ] process > Minimap2Workflow:Rma2Info      -
[-        ] process > MergeResults                   -
Error executing process > 'Demultiplex:Concatenate'

Caused by:
  Process `Demultiplex:Concatenate` terminated with an error exit status (1)

Command executed:

  cat *fastq > allfq.fastq

Command exit status:
  1

Command output:
  (empty)

Command error:
  cat: fastq: Is a directory

Work dir:
  /Users/dr/Desktop/work/ed/83f3fa22495b34f40502fbadab2da4

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`

(base) Drs-iMac:Desktop dr$ 

So do I concatenate the files?

**Reply by KFeye** [reply]
Ok sorry again..got it to work...then...


Caused by:
  Process `QFilt:Yacrd (blueplatefiles)` terminated with an error exit status (1)

Command executed:

  yacrd ueplatefiles.fas-i overlap_blueplatefiles.paf 		-o report_blueplatefiles.yacrd 		-c 4 		-n 0.4 		scrubb 		-i Filt_blueplatefiles.fastq 		-o Filt_Scrubb_blueplatefiles.fastq

Command exit status:
  1

Command output:
  (empty)

Command error:
  Error: Error in compression detection of file overlap_blueplatefiles.paf
  
  Caused by:
      File is too short, less than five bytes

Work dir:
  /Users/dr/Desktop/work/03/025c25db10611acdfb817bc3b20806

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`


Thank you again x 100000000000000000000000.  Honestly, this program is going to make this kind of sequencing so much more accessible to people like me who are more biological than bioinformatics.  Thank you for all of your help.

**Reply by iferres** [reply]
Hi, Could you show me the command you are launching?

**Reply by iferres** [reply]
Also, I'd like to know what's your background using the command line. Just to know the level of detail you may need to understand the commands.

**Reply by KFeye** [reply]
My background is purely in QIIME2 and some python in the Conda environment, and some R (ggplot2, stats, phyloseq, aldex2, etc).  The command I launched that resulted in this error is: 

nextflow run microgenlab/porefile --fq 'FASTQFilesPlate2/' --minimap2 --isDemultiplexed

I have docker and Conda and everything is up to date.  I ran the nextflow tutorial on their site and had a positive outcome.

**Comment by iferres** [comment]
Since this wasn't a bug report, I moved the issue here to discussions to follow the conversation.

**Comment by KFeye** [comment]
Thanks!  I tried to move it...but failed.

**Reply by iferres** [reply]
No problem!

**Comment by KFeye** [comment]
Hey!

My background is working with QIIME2, doing some python work in conda, and
moving files around (so super light).   I also do some work in R (phyloseq,
ggplot2, stats...)

The command line I am running is$ nextflow run microgenlab/porefile --fq
'FASTQFilesZinpro2/' --minimap2 --isDemultiplexed

N E X T F L O W  ~  version 20.10.0

I ran the sample data for Nextflow through their pipeline for a BLAST
(found on their site) and was successful.

I have docker and conda, I have everything updated completely.

**Reply by iferres** [reply]
Great!  Try:
```
nextflow run microgenlab/porefile --fq 'FASTQFilesZinpro2/*.fastq' --minimap
```
Pay attention to the glob pattern (`*.fastq`). And also, since this time you are passing the basecalled fastq files without demultiplexing (am I right?), remove the `--isDemultiplexed` flag. 

The pipeline will probably fail in the Nanoplot process. If that happens it's just that there's a problem with the conda version of nanoplot. I'm waiting for a fix from the nanoplot developers. Just add `--noNanoplot -resume`. The `-resume` flag (one dash only this time) is to use caché and avoid recomputing processes that had already finished.

**Reply by KFeye** [reply]
Hey!  Everything looked great with that code, then this happened...

Error executing process > 'Demultiplex:Porechop'

Caused by:
  Process `Demultiplex:Porechop` terminated with an error exit status (137)

Command executed:

  porechop -t 1 -i allfq.fastq -b porechop_results
  rm -f porechop_results/none.fastq

Command exit status:
  137

Command output:
  
  Loading reads
  allfq.fastq

Command error:
  .command.sh: line 2:     7 Killed                  porechop -t 1 -i allfq.fastq -b porechop_results

Work dir:
  /Users/dr.kristinafeyewaseen/Desktop/work/69/13bddb07c96f43383198760ca15713

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line

**Reply by iferres** [reply]
Error code 137 usually means that your process was killed because it used too much memory. You mentioned the possibility of running it in the cloud. A more powerful machine will probably cope with that.

**Reply by KFeye** [reply]
Hey!  So I moved to Google Cloud...I have some more errors and likely it is just something I have to shift as it goes through docker and whatnot...

kfeye@cloudshell:~ (digital-rhythm-308021)$ ./nextflow nextflow run microgenlab/porefile --fq 'blueplatefiles/*.fastq' --minimap2 
N E X T F L O W  ~  version 20.10.0
Launching `nextflow` [agitated_thompson] - revision: a1abfb246f
Script compilation error
- file : /home/kfeye/nextflow
- cause: Unexpected input: '#' @ line 2, column 1.
   #
   ^
1 error


The google cloud directions are here: https://cloud.google.com/life-sciences/docs/tutorials/nextflow#console_3

I have my data downloaded from the bin, visible, and ready to go.  I can open files and move around.  Nextflow is off.  Am I supposed to download the .sh file and adjust it as per the directions?  Does it have the same command lines? 

OR can I shrink down my data locally and process it into chunks, then merge it later?

**Reply by iferres** [reply]
You are typing `nextflow` twice: `kfeye@cloudshell:~ (digital-rhythm-308021)$ ./nextflow nextflow run ...`.
Use (from where the `nextflow` binary is located):
```
./nextflow run microgenlab/porefile --fq 'blueplatefiles/*.fastq' --minimap2
```

**Reply by KFeye** [reply]
It worked!  Their systems have weird edit requests.  I now have to upgrade my computer cloud, but it looks like we got it all solved! =)

This is great!  Thank you!  I am so excited to finally get data pushed through the system!
